### ISSUE TITLE
Error: Cannot find module 'undefined'

### ISSUE THREAD

**Issue by aaronjensen** [wallabyjs/public#252]
Every once in a while, my atom-wallaby gets into a state where React can't load anymore. If react is required, I see this in the logs:

atom-wallaby panel:
```
Error: Cannot find module 'undefined'
TypeError: 'undefined' is not a function (evaluating 'ReactElementValidator.createElement.bind(
      null,
      type
    )')
```

console:
```
Wallaby Console: Runtime error: Error: Cannot find module 'undefined'/Users/aaronjensen/.atom/packages/atom-wallaby/wallaby-atom/index.js:formatted:237 (anonymous function)
/Users/aaronjensen/.atom/packages/atom-wallaby/wallaby-atom/index.js:formatted:237Wallaby Console: Runtime error: TypeError: 'undefined' is not a function (evaluating 'ReactElementValidator.createElement.bind(/Users/aaronjensen/.atom/packages/atom-wallaby/wallaby-atom/index.js:formatted:237 (anonymous function)
/Users/aaronjensen/.atom/packages/atom-wallaby/wallaby-atom/index.js:formatted:237Wallaby Console:       null,
/Users/aaronjensen/.atom/packages/atom-wallaby/wallaby-atom/index.js:formatted:237Wallaby Console:       type
/Users/aaronjensen/.atom/packages/atom-wallaby/wallaby-atom/index.js:formatted:237Wallaby Console:     )')
/Users/aaronjensen/.atom/packages/atom-wallaby/wallaby-atom/index.js:formatted:237Wallaby Console: Runtime error: Error: Cannot find module 'undefined'
```

A stop/start seems to clear it up, but it happens fairly often.

**Comment by ArtemGovorov** [comment]
If you are using PhantomJs 1.9.8 (used in wallaby by default, but we support PhantomJs2 as well), this error `TypeError: 'undefined' is not a function (evaluating 'ReactElementValidator.createElement.bind(` happens because Phantom 1.9.8 doesn't support `Function.prototype.bind`, so a polyfill is required.

Installing a module with a polyfill (or just referencing if you already have it somewhere) and adding it to your `files` list, should do the trick and eliminate the error: 
`{pattern: 'node_modules/phantomjs-function-bind-polyfill/index.js', instrument: false},`

Let me know adding the polyfill fixes it for you.

**Comment by aaronjensen** [comment]
Ah. Well, I do run the polyfill in my specHelper.js, i'll try adding it in my wallaby.conf.js too, but it seems like the one in my specHelper.js should do. It is required via the new `webpackConfig.entryPatterns`

**Comment by aaronjensen** [comment]
btw, does the order of `entryPatterns` matter?

I had it i this order as that is what you mentioned when you added it, but it may be backwards...

```js
  webpackConfig.entryPatterns = [
    'specs/**/*.spec.js',
    'specs/specHelper.js',
  ];
```

**Comment by ArtemGovorov** [comment]
You're correct, the `specHelper.js` loading the polyfill should do, unless there's some code requiring the polyfill and executing before the `specHelper.js` gets loaded into the sandbox (it gets loaded after all files listed in your `files` list with `load` flag not set to `false`).

Order for `entryPatterns` shouldn't matter, tests (patterns from `tests` list) are always loaded after `files`. Unless there's a spec (file satisfying `specs/**/*.spec.js`) that is listed in the `files` list for some reason, but in this case there should be a warning like "some of your tests are in your files list".

Is there any specific pattern you've noticed when the error is reproduced? Like does it happen after adding new files, requiring new dependencies from files/specs, or it there's no noticeable pattern? 

Either way, try adding the polyfill to your files list and see how it goes. Let me know if it helps or doesn't, we'll see if it's an issue with the way `specHelper.js` is loaded.

**Comment by ArtemGovorov** [comment]
I think I have reproduced the issue, looking into it now.

**Comment by ArtemGovorov** [comment]
The issue was in my `entryPatterns` implementation in wallaby-webpack. Should be fixed now, please update the wallaby-webpack package to v0.0.7.

**Comment by ArtemGovorov** [comment]
Forgot to mention, no need to add the polyfill to the `files` list, should work as it was in your `specHelper`.

**Comment by aaronjensen** [comment]
Awesome, thanks for the quick fix!

**Comment by aaronjensen** [comment]
I just got this: 

```
"before each" hook: 'undefined' is not a function (evaluating 'definition[name].bind(store)')
```

It looks like the same issue :(

**Comment by aaronjensen** [comment]
hm, it seems to have happened during a wallaby update. maybe that was the issue.

**Comment by ArtemGovorov** [comment]
So after the update is finished, it works ok or still is happening?

**Comment by aaronjensen** [comment]
I restarted atom and it goes away. That said, i think I've found repros for nasty things. Wallaby doesn't like it when you remove files. I'm deleting a number of files (that would cause require breaks and test breaks) and I start seeing bind errors or other oddities.

**Comment by ArtemGovorov** [comment]
Cool, is there a way I could reproduce it in the https://github.com/aaronjensen/wallaby-line-number-repro repo? Do I just delete a few source?/test? files while wallaby is running?

**Comment by aaronjensen** [comment]
Here's one repro:

https://github.com/aaronjensen/wallaby-line-number-repro/blob/master/specs/working.spec.js#L1

**Comment by aaronjensen** [comment]
Another repro seems to be if I get the tests passing by making libf to lib and restarting wallaby a couple times, then I delete foo.js and then remove the corresponding require statement, it corrupts the cache.

**Comment by ArtemGovorov** [comment]
Great, thanks a lot for finding the way to reproduce it! Will look into it now.

**Comment by ArtemGovorov** [comment]
Ok, I can reproduce the second one 100% reliably doing exactly as you've described it.

With the first one, I can't see anything wrong, here is what I do:

![rp](https://cloud.githubusercontent.com/assets/979966/9830005/03dc10ea-5962-11e5-9bd8-57e6b169715e.gif)

What am I doing wrong?

**Comment by aaronjensen** [comment]
Hmm, I don't know. I tried it w/ just that file open and it repro'd every time. 

```
$ node -v
v0.10.26
$ npm -v
2.14.1
$ npm list | grep phant
├── karma-phantomjs-launcher@0.2.1
├─┬ phantomjs@1.9.18
├── phantomjs-function-bind-polyfill@1.0.0
```

![wallaby](https://cloud.githubusercontent.com/assets/8588/9830223/90d253fc-58da-11e5-8885-55890ab2b9d7.gif)

**Comment by ArtemGovorov** [comment]
What is your installed wallaby core version (can be viewed in Atom Developer Tools, `Cmd/Ctrl + Alt + I`)?

**Comment by aaronjensen** [comment]
```
Wallaby Console: Atom plugin v1.0.6
/Users/aaronjensen/.atom/packages/atom-wallaby/lib/wallaby.js:95 Wallaby Console: Core v1.0.103
```

**Comment by ArtemGovorov** [comment]
Oh, one more thing:

> Should be fixed now, please update the wallaby-webpack package to v0.0.7.

Did you update your wallaby-webpack package to latest?

**Comment by aaronjensen** [comment]
ah! that fixed it on this repo. Sorry for the goose chase. I believe I've seen bind errors on my real repo that was already updated, but I don't have a repro for it... I'll keep an eye out an open an issue if I figure anything else out.

**Comment by ArtemGovorov** [comment]
Awesome, thanks. I'll be pushing the fix for the second issue you've reproduced shortly.

**Comment by aaronjensen** [comment]
new repro in #260

**Comment by ArtemGovorov** [comment]
Thanks for providing the way to reproduce it! I have fixed the #260 and this one:

> Another repro seems to be if I get the tests passing by making libf to lib and restarting wallaby a couple times, then I delete foo.js and then remove the corresponding require statement, it corrupts the cache.

in core v1.0.104.

**Comment by aaronjensen** [comment]
Works like a charm. Thanks for the quick fixes.
